# Step 1: import the library
### In this step we will import the libraries we will need:
- **pandas**: to load the data, which is a library that provides high flexibility in dealing with and loading tabular data (**DataFrame**).
- **Matplotlib**: We used matplotlib library because it provides various charting tools and is the most widely used for visualization.
- **Pickle**: We use pickle so that we can work with files having the .pkl extension.
- **Sklearn.linear_model**: to use the linear model we need (*in our case, __LinearRegression__, __Ridge__, __Lasso__*).
- **Sklearn.ensemble**: to use the __RandomForestRegressor__ model.
- **Sklearn.matrics**: to use the model evaluation tools we need (*in our case, __r2_score__*).
- **Sklearn.model_selection**: to use the tools for developing and selecting models (*in our case, __train_test_split__ to split data, and __gridsearchcv__ to find the best parameter for the model*).
### The links:
- **[Scikit-learn](https://scikit-learn.org/stable/)**
- **[Pandas](https://pandas.pydata.org/)**
- **[Matplotlib](https://matplotlib.org/)**

In [1]:
import pandas as pd # to loading data
import matplotlib.pyplot as plt # for visualization
import pickle
from sklearn.linear_model import LinearRegression, Ridge, Lasso # to use the linear model we need
from sklearn.ensemble import RandomForestRegressor # to use the RandomForestRegressor
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error # to evalueation the models
from sklearn.model_selection import GridSearchCV # to find the parameter for the models
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import train_test_split # to split the data

# Step 2: load the data
### In this step we will load tabular data from boston.csv file using pandas which is table with this features:
- `CRIM`: per capita crime rate by town
- `ZN`: proportion of residential land zoned for lots over 25,000 sq.ft
- `INDUS`: proportion of non-retail business acres per town.
- `CHAS`: Charles River dummy variable (1 if tract bounds river; 0 otherwise).
- `NOX`: nitric oxides concentration (parts per 10 million) [parts/10M].
- `RM`: average number of rooms per dwelling.
- `AGE`: proportion of owner-occupied units built prior to 1940.
- `DIS`: weighted distances to five Boston employment centres.
- `RAD`: index of accessibility to radial highways.
- `TAX`: full-value property-tax rate per $10,000 [$/10k].
- `PTRATIO`: pupil-teacher ratio by town.
- `B`: The result of the equation B=1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town (*this column will be removed because it expresses racial variables, which is considered illogical and racist*).
- `LSTAT`: % lower status of the population.
### And the target:
- `MEDV`: the price of the houce.

In [2]:
data = pd.read_csv("boston.csv") # loading data from boston.csv file
data

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


In [4]:
data.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

# Step 3: drop and remove the are not afraid of it
### In this step we remove this features (*columns*):
 1) **CHAS**: Because it is correlation with price is weak. (*We will know this in the next codes.*)
 2) **B**: Becouse it is illogical and racist.

In [5]:
correlation = data.corr().loc[["MEDV"]] # find the correlation 
correlation

CRIM        ZN     INDUS     CHAS       NOX       RM       AGE  \
MEDV -0.388305  0.360445 -0.483725  0.17526 -0.427321  0.69536 -0.376955   

           DIS       RAD       TAX   PTRATIO         B     LSTAT  MEDV  
MEDV  0.249929 -0.381626 -0.468536 -0.507787  0.333461 -0.737663   1.0

In [6]:
data.drop(["CHAS", "B"], axis=1, inplace=True)

In [7]:
X = data.drop("MEDV", axis=1) # this mean X = All features except MEDV
y = data["MEDV"] # and y = the target MEDV

# Step 4: spliting the data
### To train our model, we need to split the data into training and testing sets.

- **X**: The features (independent variables) used to make predictions.
- **y**: The target variable (what we want to predict).
```python
X = data.drop("MEDV", axis=1)
y = data["MEDV"]
```

- **X_train**: 80% of `X`, used to train the model.
- **X_test**: 20% of `X`, used to evaluate the model.
- **y_train**: Corresponding 80% of `y`, for training.
- **y_test**: Corresponding 20% of `y`, for evaluation.

### We use `train_test_split` from `sklearn` to split the data:
```python
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=33) # spliting the data

In [12]:
models = {
    'Linear Regression': {
        "model": LinearRegression(),
        "parameters": {
            'fit_intercept': [True, False],
            'copy_X': [True, False],
            'n_jobs': [None, -1],  # يمكنك تحديد عدد الأنوية هنا'positive': [True, False]  # لتقييد المعاملات بالقيم الموجبة فقط
        }
    },
    'Ridge Regression': {
        "model": Ridge(),
        "parameters": {
            'alpha': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100],
            'solver': ['auto', 'svd', 'cholesky', 'lsqr', 'sparse_cg', 'sag', 'saga'],
            'fit_intercept': [True, False],
            'tol': [1e-4, 1e-3, 1e-2]
        }
    },
    'Lasso Regression': {
        "model": Lasso(),
        "parameters": {
            'alpha': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100],
            'max_iter': [1000, 5000, 10000],
            'selection': ['cyclic', 'random'],
            'tol': [1e-4, 1e-3, 1e-2],
            'fit_intercept': [True, False],
            'warm_start': [True, False]
        }
    },
    'Random Forest': {
        "model": RandomForestRegressor(warm_start=True),
        "parameters": {
            'n_estimators': [10, 50, 100, 200, 500],
            'criterion': ['squared_error', 'absolute_error'],  # للتصنيف استخدم ['gini', 'entropy']
            'max_depth': [None, 10, 20, 30, 50],
            'bootstrap': [True],
            'max_features': ['sqrt', 'log2', None],
            'oob_score': [True, False],
            'n_jobs': [-1],  # استخدام جميع الأنوية المتاحة
            'verbose': [0, 1, 2],
            'random_state': [42]
        }
    },
}
models

{'Linear Regression': {'model': LinearRegression(),
  'parameters': {'fit_intercept': [True, False],
   'copy_X': [True, False],
   'n_jobs': [None, -1]}},
 'Ridge Regression': {'model': Ridge(),
  'parameters': {'alpha': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100],
   'solver': ['auto', 'svd', 'cholesky', 'lsqr', 'sparse_cg', 'sag', 'saga'],
   'fit_intercept': [True, False],
   'tol': [0.0001, 0.001, 0.01]}},
 'Lasso Regression': {'model': Lasso(),
  'parameters': {'alpha': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100],
   'max_iter': [1000, 5000, 10000],
   'selection': ['cyclic', 'random'],
   'tol': [0.0001, 0.001, 0.01],
   'fit_intercept': [True, False],
   'warm_start': [True, False]}},
 'Random Forest': {'model': RandomForestRegressor(warm_start=True),
  'parameters': {'n_estimators': [10, 50, 100, 200, 500],
   'criterion': ['squared_error', 'absolute_error'],
   'max_depth': [None, 10, 20, 30, 50],
   'bootstrap': [True],
   'max_features': ['sqrt', 'log2', None],
   'oob_score': [True,

# Step 5: try the models and choose the best one
### In this step we will:
1. Get the best parameters to tring the models.
2. Train each models using **80%** of the data: `X_train` and `y_train`.
3. Testing the models using **20%** of the data: `X_test` and `y_test`.

In [13]:
def best_parameters(models, X_train, y_train):
    best_params_models = {}
    for name, details in models.items():
        if name == "Random Forest":
            randomized_search = RandomizedSearchCV(error_score="raise", estimator=details["model"], param_distributions=details["parameters"], cv=5, n_jobs=-1, verbose=2)
            randomized_search.fit(X_train, y_train)
            print(f"Best parameters found for {name}: {randomized_search.best_params_}")
            best_params_models[name] = {
                "model": randomized_search.best_estimator_,
                "parameters": randomized_search.best_params_
            }
        else:
            grid_search = GridSearchCV(error_score="raise", estimator=details["model"], param_grid=details["parameters"], cv=5, n_jobs=-1, verbose=2)
            grid_search.fit(X_train, y_train)
            print(f"Best parameters found for {name}: {grid_search.best_params_}")
            # the best parameters
            best_params_models[name] = {
                "model": grid_search.best_estimator_,
                "parameters": grid_search.best_params_
            }
    return best_params_models

In [14]:
best_params_models = best_parameters(models, X_train, y_train)
best_params_models

Fitting 5 folds for each of 8 candidates, totalling 40 fits
Best parameters found for Linear Regression: {'copy_X': True, 'fit_intercept': True, 'n_jobs': None}
Fitting 5 folds for each of 294 candidates, totalling 1470 fits
Best parameters found for Ridge Regression: {'alpha': 1, 'fit_intercept': True, 'solver': 'svd', 'tol': 0.0001}
Fitting 5 folds for each of 504 candidates, totalling 2520 fits
Best parameters found for Lasso Regression: {'alpha': 0.01, 'fit_intercept': True, 'max_iter': 5000, 'selection': 'random', 'tol': 0.001, 'warm_start': False}
Fitting 5 folds for each of 10 candidates, totalling 50 fits
Best parameters found for Random Forest: {'verbose': 0, 'random_state': 42, 'oob_score': True, 'n_jobs': -1, 'n_estimators': 500, 'max_features': None, 'max_depth': 30, 'criterion': 'squared_error', 'bootstrap': True}


{'Linear Regression': {'model': LinearRegression(),
  'parameters': {'copy_X': True, 'fit_intercept': True, 'n_jobs': None}},
 'Ridge Regression': {'model': Ridge(alpha=1, solver='svd'),
  'parameters': {'alpha': 1,
   'fit_intercept': True,
   'solver': 'svd',
   'tol': 0.0001}},
 'Lasso Regression': {'model': Lasso(alpha=0.01, max_iter=5000, selection='random', tol=0.001),
  'parameters': {'alpha': 0.01,
   'fit_intercept': True,
   'max_iter': 5000,
   'selection': 'random',
   'tol': 0.001,
   'warm_start': False}},
 'Random Forest': {'model': RandomForestRegressor(max_depth=30, max_features=None, n_estimators=500,
                        n_jobs=-1, oob_score=True, random_state=42,
                        warm_start=True),
  'parameters': {'verbose': 0,
   'random_state': 42,
   'oob_score': True,
   'n_jobs': -1,
   'n_estimators': 500,
   'max_features': None,
   'max_depth': 30,
   'criterion': 'squared_error',
   'bootstrap': True}}}

# Step 6: get the results
Evaluation the models using `r2_score` (*The higher the __r2_score number__, the better.*), to choose the best model

In [17]:
def model_trying(models_details, X_test, y_test):
    result = {}
    for name, model in models_details.items():
        y_pred = model["model"].predict(X_test)
        result[name] = {
            "model": model["model"],
            "R²": r2_score(y_test, y_pred),
            "RMSE": mean_squared_error(y_test, y_pred),
            "MAE": mean_absolute_error(y_test, y_pred)
        }
    return result

In [18]:
models_details = model_trying(best_params_models, X_test, y_test)
models_details

{'Linear Regression': {'model': LinearRegression(),
  'R²': 0.682204188548306,
  'RMSE': 22.765134036498836,
  'MAE': 3.686644639104774},
 'Ridge Regression': {'model': Ridge(alpha=1, solver='svd'),
  'R²': 0.6720123566569238,
  'RMSE': 23.495220496810933,
  'MAE': 3.73248469819937},
 'Lasso Regression': {'model': Lasso(alpha=0.01, max_iter=5000, selection='random', tol=0.001),
  'R²': 0.678302118086267,
  'RMSE': 23.044656779993783,
  'MAE': 3.6846024102558808},
 'Random Forest': {'model': RandomForestRegressor(max_depth=30, max_features=None, n_estimators=500,
                        n_jobs=-1, oob_score=True, random_state=42,
                        warm_start=True),
  'R²': 0.8267259143597692,
  'RMSE': 12.412397025098072,
  'MAE': 2.2050901960784315}}

# Step 7: choose the best model
### What we notice:
This column gives the r2 score for each of the following models:

<img src="MAE result.png" width="500" align="right" style="margin: 10px">
1. **`RandomForestRegression`**: **R²** is 0.8267259143597692, **RMSE** is 12.412397025098072, **MAE** is 2.2050901960784315, is the best model.
2. **`LinearRegression`**: **R²** is 0.682204188548306, **RMSE** is 22.765134036498836, **MAE** is 3.686644639104774.
3. **`Lasso`**: **R²** is 0.678302118086267, **RMSE** is 23.044656779993783, **MAE** is 3.6846024102558808.
4. **`Ridge`**: **R²** is 0.6720123566569238, **RMSE** is 23.495220496810933, **MAE** is 3.73248469819937.

<img src="R² result.png" width="500" style="margin: 10px">
- The **LinearRegression**, **Ridge**, **Lasso** models were close together.
- The **RandomForestRegression** model outperforms them by a large margin.
you can see this in this chart...
<img src="RMSE result.png" width="500"style="margin: 10px">

### What we conclude:
This means that the relationship between the features is non-linear, since the **RandomForestRegression** model was able to achieve high accuracy, and it is good at predicting the target that has non-linear features.


# Step 8: save the models:
### we will use th pickle library to save th models
in models_detail.pkl file, so we can use it again

In [28]:
with open("models_detail.pkl", "wb") as file:
    pickle.dump([best_params_models, models_details], file)

### (*We can retrieve the model details at any time using the following code.*):
```python
with open("model_details.pkl", "rb") as file:
    RandomForestRegressor_details = pickle.load(file)